ngx_http_sxg_filter_module.c

```diff
@@ -303,38 +333,123 @@ static ngx_str_t extract_angled_url(char* str, size_t len) {
   return (ngx_str_t){.data = NULL, .len = 0};
 }
 
-static void extract_preaload_url_list(ngx_str_t* link, ngx_array_t* const dst,
-                                      ngx_http_request_t* r) {
+// TODO(kumagi): Complex logic should be migrated to ngx_sxg_utils.c.
+// Extracts URL list `dst` from `link` like </foo.js>;rel="preload";as="script"
+// Returns length of non-preload header string.
+// If `dst` is NULL, it just calculate the length of estimatec
+// non_preload_headers.
+static size_t extract_preload_url_list(ngx_str_t* link, ngx_array_t* const dst,
+                                       ngx_str_t* non_preload_headers,
+                                       ngx_http_request_t* r) {
   char* str = (char*)link->data;
-  char* const end = str + link->len;
+  char* end = str + link->len;
+  size_t non_preload_headers_len = 0;
   while (str < end) {
     const size_t tail = get_term_length(str, end - str, ',', "<>");
     ngx_str_t url = extract_angled_url(str, tail);
     char* param = str;
+    bool found = false;
+    ngx_subresource_t* new_preload = NULL;
+    ngx_str_t as = ngx_null_string;
+
     while (param < str + tail) {
       const size_t rest = str + tail - param;
       size_t param_tail = get_term_length(param, rest, ';', "<>");
+
       if (param_is_preload(param, param_tail)) {
-        ngx_str_t* const new_url = ngx_array_push(dst);
-        *new_url = url;
-        break;
+        if (dst != NULL) {
+          new_preload = ngx_array_push(dst);
+          new_preload->url.data = ngx_palloc(r->pool, url.len);
+          new_preload->url.len = url.len;
+          ngx_memcpy(new_preload->url.data, url.data, url.len);
+          while (new_preload->url.data[new_preload->url.len - 1] == ' ') {
+            new_preload->url.len--;
+          }
+          if (as.data != NULL) {
+            new_preload->as = as;
+          }
+        }
+        found = true;
+      }
+      if (param_is_as(param, param_tail, (const char**)&as.data, &as.len)) {
```

Comment:
Should this be inside the above `if (param_is_preload(...))`? Otherwise, I think there is a chance it would set `as` on the wrong `new_preload`, e.g. for a header like:

```
Link: <foo>;rel=preload,<bar>;as=image
```